specs/derivation.md

```diff
@@ -0,0 +1,537 @@
+
+By default, ethereum calldata is supported as data-availability source.
+
+Transactions of the queried block are traversed, and filtered by receiver and sender:
+- The transaction receiver MUST be the sequencer inbox address
+- The transaction must be signed by a recognized batch submitter account
+
+
+### ChannelBank
+
+#### Constants
+
+```text
+CHANNEL_TIMEOUT = 600   # time in seconds until a channel is pruned from the channel bank
+MAX_CHANNEL_BANK_SIZE = 100_000_000  # max size of the combined channels in the channel bank
+FRAME_OVERHEAD = 200    # space each frame in a channel takes, even without frame contents
+```
+
+The channel bank processes L1 transactions and buffers the encoded frames to be read.
+
+Only future frames are buffered. If there is a full frame ready to read, it MUST be read first before a new L1 transaction can be processed.
+
+The channel bank does not handle reorgs, see the recovering-from-L1-reorgs section for information how these are handled.
+
+The channel bank tracks:
+
+- The collection of open channels, ordered in a FIFO queue. The first seen channel has priority.
+- `currentL1Origin`: the current L1 block, updated every time a new block is appended to the L1 chain, before any of the data of the block is processed in the bank.
+
+Channel IDs are random `bytes32` value with timestamp.
+In this may become a hash of the sender and a channel nonce or other similar information to separate control of channels between different submitters.
+
+When `channel.time + CHANNEL_TIMEOUT < currentL1Origin.timestamp` an open channel is removed, cannot be re-instantiated anymore, and cannot be read from anymore.
+When a frame is missing this limits the amount of time that later frames pile up, preventing potential out-of-memory DoS and enabling a new channel to be read.
+
+A channel in the bank tracks:
+
+- its ID: random `bytes32` and timestamp that the first frame, part of the identity of the channel.
+- `progress`: the frame number of the current frame to read, which may not be buffered yet. Initially 0.
+- `ends_at`: the frame number of the frame that closed the channel. Initially set to a max uint64 integer value, mark there is no end.
+- The existing collection of frames tagged with their respective origin information. Each entry is keyed by frame number.
+- `size`: the amount of data the open channel has buffered.
+- Its channel ID and index in the priority queue.
+
+The channel bank processes L1 data as follows:
+
+1. Read the version number of the L1 data, check if it is 0, the data is invalid if not.
+2. Determine the `total_size` of the bank by summing the `size` of all channels in the bank.
+   For as long as `total_size > MAX_CHANNEL_BANK_SIZE`    prune the first-seen channel (i.e. lowest `time`) and reduce `total_size` by that channel `size` .
+   The first-seen channel is pruned first, since this is the channel that fails to be read before the bank became this large.
```

Comment:
I think we should prune the last seen to enable the node to keep processing the open channel. There's also an argument that we should reopen future channels after we've cleared the current channel.